File: colorpickerview/src/main/java/me/daemon/colorpicker/ColorPickerView.java
package me.daemon.colorpicker;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.PointF;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewDebug;
import android.view.ViewParent;

import java.util.ArrayList;
import java.util.List;

import me.daemon.colorpicker.painter.DefaultIndicatorPainter;
import me.daemon.colorpicker.painter.DefaultPalettePainter;
import me.daemon.colorpicker.painter.IndicatorPainter;
import me.daemon.colorpicker.painter.PalettePainter;

/**
 * color picker view
 * 颜色选择器View
 *
 * @author daemon
 * @since 2019-01-27 18:04
 */
public class ColorPickerView extends View implements ColorObservable {

    /**
     * initial color
     * 初始颜色值
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    private int initialColor;

    /**
     * padding of palette
     * 调色板留白
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    private int palettePadding;

    /**
     * whether automatic measure this ColorPickerView as square,
     * if true then minimum size of width and height will used as
     * measured width and height
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    private boolean square = false;

    private final PointF currentPoint;

    private int paletteCenterX;
    private int paletteCenterY;

    private final BrightnessProvider defaultBrightnessProvider = new BrightnessProvider() {
        @Override
        public float getBrightness() {
            return 1.0f;
        }
    };

    private BrightnessProvider brightnessProvider;

    private final PalettePainter defaultPalettePainter = new DefaultPalettePainter();

    private PalettePainter palettePainter = null;

    private final IndicatorPainter defaultIndicatorPainter = new DefaultIndicatorPainter();

    private IndicatorPainter indicatorPainter = null;

    private final List<ColorObserver> observers = new ArrayList<>();

    private int color;

    private boolean isChanging = false;

    private boolean disallowInterceptTouchEven = false;

    public ColorPickerView(Context context) {
        this(context, null);
    }

    public ColorPickerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ColorPickerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        currentPoint = new PointF();

        final TypedArray t = context.obtainStyledAttributes(attrs, R.styleable.ColorPickerView);

        try {
            final int palettePadding = (int) t.getDimension(R.styleable.ColorPickerView_palettePadding, 0);
            setPalettePadding(palettePadding);

            final int initialColor = t.getColor(R.styleable.ColorPickerView_initialColor, Color.BLACK);
            setInitialColor(initialColor);
            setColorInternal(initialColor, false);

            final boolean disallowInterceptTouchEven = t.getBoolean(
                    R.styleable.ColorPickerView_disallowInterceptTouchEvent,
                    false);
            setDisallowInterceptTouchEven(disallowInterceptTouchEven);
        } finally {
            t.recycle();
        }

    }

    /**
     * set padding of palette
     * 设置调色板留白
     *
     * @param palettePadding 调色板留白
     */
    public void setPalettePadding(final int palettePadding) {
        if (this.palettePadding != palettePadding) {
            this.palettePadding = palettePadding;

            invalidate();
        }
    }

    public int getPalettePadding() {
        return palettePadding;
    }

    /**
     * set initial color
     * 设置初始颜色
     *
     * @param initialColor 初始颜色
     */
    public void setInitialColor(final int initialColor) {
        if (this.initialColor != initialColor) {
            this.initialColor = initialColor;

            invalidate();
        }
    }

    public int getInitialColor() {
        return initialColor;
    }

    /**
     * set square mode
     * <p>
     * 设置正方形模式
     *
     * @param square whether use square mode
     */
    public void setSquare(final boolean square) {
        if (this.square != square) {
            this.square = square;

            invalidate();
        }
    }

    public boolean getSquare() {
        return square;
    }

    /**
     * set current picked color
     * 设置选中颜色值
     *
     * @param color 颜色值
     */
    public void setColor(final int color) {
        setColorInternal(color, true);
    }

    /**
     * resolve touch conflict with it's parent and ancestors<br/><br/>
     * <p>
     * if disallow is set to be true, {@link ColorPickerView}  will call
     * it's {@link ViewParent#getParent() parent}'s
     * {@link ViewParent#requestDisallowInterceptTouchEvent(boolean)}
     * when received {@link MotionEvent#ACTION_DOWN} event in
     * {@link #onTouchEvent(MotionEvent)}
     * to disallow it's parent and ancestors to intercept touch event,
     * and restore when received {@link MotionEvent#ACTION_UP} event
     * <br/><br/>
     * 解决{@link ColorPickerView}与其父View或祖先View的触摸事件冲突<br/><br/>
     * <p>
     * disallow设置为true时，在{@link #onTouchEvent(MotionEvent)}
     * 方法中接收到{@link MotionEvent#ACTION_DOWN} 触摸事件时会调用
     * {@link ViewParent#getParent() 父View}的
     * {@link ViewParent#requestDisallowInterceptTouchEvent(boolean)}方法来紧张父View
     * 及祖先View拦截触摸事件，并在收到{@link MotionEvent#ACTION_UP}触摸事件时恢复
     *
     * @param disallow whether to disallow it's parent and ancestors to intercept touch event<br/>
     *                 是否禁止父View或祖先View拦截触摸事件冲突
     */
    public void setDisallowInterceptTouchEven(final boolean disallow) {
        this.disallowInterceptTouchEven = disallow;
    }

    private void setColorInternal(final int color, final boolean notify) {
        this.color = color;

        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        float r = hsv[1] * getRadius();
        float radian = (float) (hsv[0] / 180f * Math.PI);
        updateIndicator((float) (r * Math.cos(radian) + paletteCenterX), (float) (-r * Math.sin(radian) + paletteCenterY));

        if (notify) {
            notifyObservers(color);
        }

        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (!square) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
            return;
        }

        final int width = MeasureSpec.getSize(widthMeasureSpec);
        final int height = MeasureSpec.getSize(heightMeasureSpec);

        final int size = Math.min(width, height);

        super.onMeasure(
                MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY)
        );
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        paletteCenterX = w / 2;
        paletteCenterY = h / 2;

        getPalettePainter().onSizeChanged(w, h, getRadius(), paletteCenterX, paletteCenterY);

        setColorInternal(color, false);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        final int radius = getRadius();

        getPalettePainter().drawPalette(
                this,
                canvas,
                radius, paletteCenterX,
                paletteCenterY
        );

        drawIndicator(canvas);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (!isEnabled()) return super.onTouchEvent(event);

        final boolean clickable = isClickable()
                || isLongClickable()
                || Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && isContextClickable();
        if (!clickable) return super.onTouchEvent(event);

        final float x = event.getX();
        final float y = event.getY();

        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                if (disallowInterceptTouchEven) {
                    // resolve touch conflicts
                    final ViewParent parent = getParent();
                    if (parent != null) {
                        parent.requestDisallowInterceptTouchEvent(true);
                    }
                }

                isChanging = true;
                update(x, y);
                return true;

            case MotionEvent.ACTION_MOVE:
                update(x, y);
                break;

            case MotionEvent.ACTION_UP:
                isChanging = false;
                update(x, y);

                performClick();

                if (disallowInterceptTouchEven) {
                    // resolve touch conflicts
                    final ViewParent parent = getParent();
                    if (parent != null) {
                        parent.requestDisallowInterceptTouchEvent(false);
                    }
                }
                break;
        }
        return super.onTouchEvent(event);
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    private static class SavedState extends View.BaseSavedState {

        private int color;

        public SavedState(Parcel source) {
            super(source);
            color = source.readInt();
        }

        @TargetApi(Build.VERSION_CODES.N)
        public SavedState(Parcel source, ClassLoader loader) {
            super(source, loader);
        }

        public SavedState(Parcelable superState) {
            super(superState);
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeInt(color);
        }

        public static final Parcelable.Creator<SavedState> CREATOR
                = new Parcelable.ClassLoaderCreator<SavedState>() {

            @Override
            public SavedState createFromParcel(Parcel source) {
                return new SavedState(source);
            }

            @Override
            public SavedState createFromParcel(Parcel source, ClassLoader loader) {
                return new SavedState(source, loader);
            }

            @Override
            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final SavedState savedState = new SavedState(super.onSaveInstanceState());

        savedState.color = color;

        return savedState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state instanceof SavedState) {
            final SavedState savedState = (SavedState) state;
            super.onRestoreInstanceState(savedState.getSuperState());

            setColor(savedState.color);

        } else {
            super.onRestoreInstanceState(state);
        }
    }

    protected void drawIndicator(final Canvas canvas) {
        final IndicatorPainter provider = indicatorPainter != null ? indicatorPainter : defaultIndicatorPainter;
        provider.drawIndicator(this, canvas, currentPoint, color, isChanging);
    }

    private int getRadius() {
        final int width = getMeasuredWidth() - getPaddingLeft() - getPaddingRight() - palettePadding * 2;
        final int height = getMeasuredHeight() - getPaddingTop() - getPaddingBottom() - palettePadding * 2;
        final int limit = Math.max(Math.min(width, height), 0);

        return limit / 2;
    }

    private void update(final float eventX, final float eventY) {
        final int color = getColorAtPoint(eventX, eventY);
        setColorInternal(color, true);
    }

    private void updateIndicator(final float eventX, final float eventY) {
        float x = eventX - paletteCenterX;
        float y = eventY - paletteCenterY;
        double r = Math.sqrt(x * x + y * y);

        final int radius = getRadius();
        if (r > radius) {
            x *= radius / r;
            y *= radius / r;
        }
        currentPoint.x = x + paletteCenterX;
        currentPoint.y = y + paletteCenterY;

        invalidate();

    }

    /**
     * set custom palette painter, using {@link #defaultPalettePainter} as default
     * palette painter
     * <p>
     * 设置自定义调色板绘制器，默认使用{@link #defaultPalettePainter}
     *
     * @param palettePainter custom palette painter
     *                       调色板绘制器
     */
    public void setPalettePainter(final PalettePainter palettePainter) {
        this.palettePainter = palettePainter;
    }

    /**
     * set custom indicator painter, using {@link #defaultIndicatorPainter}
     * if custom indicator not been set
     * <p>
     * 设置自定义指示器绘制器，如果没有设置则使用{@link #defaultIndicatorPainter 默认绘制器}
     *
     * @param indicatorPainter custom indicator painter
     */
    public void setIndicatorPainter(final IndicatorPainter indicatorPainter) {
        this.indicatorPainter = indicatorPainter;
    }

    /**
     * set custom brightness provider, {@link #defaultBrightnessProvider} is
     * used if custom brightness provider is not been set
     * <p>
     * 设置自定义透明度提供器，为空则使用{@link #defaultBrightnessProvider 默认透明度提供器}
     *
     * @param brightnessProvider custom brightness provider
     */
    public void setBrightnessProvider(final BrightnessProvider brightnessProvider) {
        this.brightnessProvider = brightnessProvider;
    }

    @Override
    public void subscribe(ColorObserver observer) {
        if (observer != null) {
            observers.add(observer);
        }
    }

    @Override
    public void unsubscribe(ColorObserver observer) {
        if (observer != null) {
            observers.remove(observer);
        }
    }

    @Override
    public int getColor() {
        return color;
    }

    private int getColorAtPoint(float eventX, float eventY) {
        float x = eventX - paletteCenterX;
        float y = eventY - paletteCenterY;
        double r = Math.sqrt(x * x + y * y);
        float[] hsv = {0, 0, 0};

        final BrightnessProvider bp = brightnessProvider != null ?
                brightnessProvider : defaultBrightnessProvider;
        final float brightness = bp.getBrightness();

        hsv[0] = (float) (Math.atan2(y, -x) / Math.PI * 180f) + 180;
        hsv[1] = Math.max(0f, Math.min(1f, (float) (r / getRadius())));
        hsv[2] = Math.max(0f, Math.min(1f, brightness));
        return Color.HSVToColor(hsv);
    }

    private void notifyObservers(final int color) {
        for (ColorObserver observer : observers) {
            observer.onColor(color);
        }
    }

    private PalettePainter getPalettePainter() {
        return palettePainter != null ?
                palettePainter : defaultPalettePainter;
    }
}

File: colorpickerview/src/main/kotlin/me/daemon/colorpicker/ColorPickerView.kt
package me.daemon.colorpicker

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.os.Parcelable
import android.util.AttributeSet
import android.view.*
import me.daemon.colorpicker.internal.ColorCenter
import me.daemon.colorpicker.painter.IAlphaPainter
import me.daemon.colorpicker.painter.IBrightnessPainter
import me.daemon.colorpicker.painter.IPalettePainter
import me.daemon.colorpicker.painter.impl.DefaultAlphaPainter
import me.daemon.colorpicker.painter.impl.DefaultBrightnessPainter
import me.daemon.colorpicker.view.AlphaView
import me.daemon.colorpicker.view.BrightnessView
import me.daemon.colorpicker.view.PaletteView

/**
 * color picker view
 *
 * 颜色选择器View
 *
 * [GitHub](https://github.com/daemon369/ColorPickerView)
 *
 * [jcenter](https://bintray.com/beta/#/daemon336699/maven/colorpickerview?tab=overview)
 *
 * @author daemon
 * @since 2019-02-26 21:53
 */
class ColorPickerView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : ViewGroup(context, attrs, defStyleAttr), ColorObservable {

    /**
     * radius of palette
     *
     * 调色板半径
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var paletteRadius: Int = 0
        /**
         * set radius of palette
         *
         * 设置调色板半径
         */
        set(paletteRadius) {
            if (field != paletteRadius) {
                field = paletteRadius

                invalidate()
            }
        }

    /**
     * palette gravity
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var paletteGravity: Gravity = Gravity.CENTER
        set(paletteGravity) {
            if (field != paletteGravity) {
                field = paletteGravity

                invalidate()
            }
        }

    /**
     * palette center offset x
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var paletteOffsetX: Int = 0
        set(paletteOffsetX) {
            if (field != paletteOffsetX) {
                field = paletteOffsetX

                invalidate()
            }
        }

    /**
     * palette center offset y
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var paletteOffsetY: Int = 0
        set(paletteOffsetY) {
            if (field != paletteOffsetY) {
                field = paletteOffsetY

                invalidate()
            }
        }

    private val colorCenter = ColorCenter()

    private val paletteView = PaletteView(context, attrs).apply {
        setColorCenter(colorCenter)
    }

    private val brightnessView = BrightnessView(context, attrs).apply {
        setColorCenter(colorCenter)
        painter = DefaultBrightnessPainter()
    }

    private val alphaView = AlphaView(context, attrs).apply {
        setColorCenter(colorCenter)
        painter = DefaultAlphaPainter()
    }

    private var isAddingInternal = false

    private var disallowInterceptTouchEvent = false

    /**
     * enable brightness view if `brightnessEnable` is true
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessEnable: Boolean = true
        set(brightnessEnable) {
            if (field != brightnessEnable) {
                field = brightnessEnable

                invalidate()
            }
        }

    /**
     * width of brightness view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessWidth: Int = 0
        set(brightnessWidth) {
            if (field != brightnessWidth) {
                field = brightnessWidth

                invalidate()
            }
        }

    /**
     * height of brighness view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessHeight: Int = 0
        set(brightnessHeight) {
            if (field != brightnessHeight) {
                field = brightnessHeight

                invalidate()
            }
        }

    /**
     * gravity of brightness view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessGravity: Gravity = Gravity.CENTER_BOTTOM
        set(brightnessGravity) {
            if (field != brightnessGravity) {
                field = brightnessGravity

                invalidate()
            }
        }

    /**
     * brightness view offset x
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessOffsetX: Int = 0
        set(brightnessOffsetX) {
            if (field != brightnessOffsetX) {
                field = brightnessOffsetX

                invalidate()
            }
        }

    /**
     * brightness view offset y
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var brightnessOffsetY: Int = 0
        set(brightnessOffsetY) {
            if (field != brightnessOffsetY) {
                field = brightnessOffsetY

                invalidate()
            }
        }

    /**
     * enable alpha view if `alphaEnable` is true
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaEnable: Boolean = true
        set(alphaEnable) {
            if (field != alphaEnable) {
                field = alphaEnable

                invalidate()
            }
        }

    /**
     * width of alpha view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaWidth: Int = 0
        set(alphaWidth) {
            if (field != alphaWidth) {
                field = alphaWidth

                invalidate()
            }
        }

    /**
     * height of alpha view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaHeight: Int = 0
        set(alphaHeight) {
            if (field != alphaHeight) {
                field = alphaHeight

                invalidate()
            }
        }

    /**
     * gravity of alpha view
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaGravity: Gravity = Gravity.CENTER_BOTTOM
        set(alphaGravity) {
            if (field != alphaGravity) {
                field = alphaGravity

                invalidate()
            }
        }

    /**
     * alpha view offset x
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaOffsetX: Int = 0
        set(alphaOffsetX) {
            if (field != alphaOffsetX) {
                field = alphaOffsetX

                invalidate()
            }
        }

    /**
     * alpha view offset y
     */
    @ViewDebug.ExportedProperty(category = "daemon")
    var alphaOffsetY: Int = 0
        set(alphaOffsetY) {
            if (field != alphaOffsetY) {
                field = alphaOffsetY

                invalidate()
            }
        }

    init {
        @SuppressLint("CustomViewStyleable")
        val t = context.obtainStyledAttributes(attrs, R.styleable.DaemonCpColorPickerView)

        try {
            // base
            val initialColor = t.getColor(R.styleable.DaemonCpColorPickerView_daemon_cp_initialColor, Color.BLACK)
            setColor(initialColor)

            val disallowInterceptTouchEvent = t.getBoolean(
                    R.styleable.DaemonCpColorPickerView_daemon_cp_disallowInterceptTouchEvent,
                    false)
            setDisallowInterceptTouchEvent(disallowInterceptTouchEvent)

            // palette
            paletteRadius = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_paletteRadius, 0f).toInt()
            val paletteGravityInt = t.getInt(R.styleable.DaemonCpColorPickerView_daemon_cp_paletteGravity, 0)
            paletteGravity = if (paletteGravityInt == 0) {
                Gravity.CENTER
            } else {
                Gravity.from(paletteGravityInt)
            }

            paletteOffsetX = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_paletteOffsetX, 0f).toInt()
            paletteOffsetY = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_paletteOffsetY, 0f).toInt()

            // brightness
            brightnessEnable = t.getBoolean(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessEnable, true)
            brightnessWidth = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessWidth, 0f).toInt()
            brightnessHeight = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessHeight, 0f).toInt()
            val brightnessGravityInt = t.getInt(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessGravity, 0)
            brightnessGravity = if (brightnessGravityInt == 0) {
                Gravity.CENTER
            } else {
                Gravity.from(brightnessGravityInt)
            }
            brightnessOffsetX = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessOffsetX, 0f).toInt()
            brightnessOffsetY = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_brightnessOffsetY, 0f).toInt()

            // alpha
            alphaEnable = t.getBoolean(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaEnable, true)
            alphaWidth = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaWidth, 0f).toInt()
            alphaHeight = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaHeight, 0f).toInt()
            val alphaGravityInt = t.getInt(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaGravity, 0)
            alphaGravity = if (alphaGravityInt == 0) {
                Gravity.CENTER
            } else {
                Gravity.from(alphaGravityInt)
            }
            alphaOffsetX = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaOffsetX, 0f).toInt()
            alphaOffsetY = t.getDimension(R.styleable.DaemonCpColorPickerView_daemon_cp_alphaOffsetY, 0f).toInt()
        } finally {
            t.recycle()
        }

        addViewInternal(paletteView)
        addViewInternal(brightnessView)
        addViewInternal(alphaView)

        clipChildren = false
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {

        measureChild(
                paletteView,
                MeasureSpec.makeMeasureSpec(paletteRadius * 2, MeasureSpec.EXACTLY),
                MeasureSpec.makeMeasureSpec(paletteRadius * 2, MeasureSpec.EXACTLY)
        )

        if (brightnessEnable) {
            measureChild(
                    brightnessView,
                    MeasureSpec.makeMeasureSpec(brightnessWidth, MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(brightnessHeight, MeasureSpec.EXACTLY)
            )
        }

        if (alphaEnable) {
            measureChild(
                    alphaView,
                    MeasureSpec.makeMeasureSpec(alphaWidth, MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(alphaHeight, MeasureSpec.EXACTLY)
            )
        }

        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        layout(
                paletteView,
                paletteGravity,
                paletteOffsetX,
                paletteOffsetY
        )

        if (brightnessEnable) {
            layout(
                    brightnessView,
                    brightnessGravity,
                    brightnessOffsetX,
                    brightnessOffsetY
            )
        }

        if (alphaEnable) {
            layout(
                    alphaView,
                    alphaGravity,
                    alphaOffsetX,
                    alphaOffsetY
            )
        }
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        when (ev.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                if (disallowInterceptTouchEvent) {
                    // resolve touch conflicts
                    parent?.requestDisallowInterceptTouchEvent(true)
                }
            }

            MotionEvent.ACTION_UP -> {
                if (disallowInterceptTouchEvent) {
                    // resolve touch conflicts
                    parent?.requestDisallowInterceptTouchEvent(false)
                }
            }
        }

        return super.onInterceptTouchEvent(ev)
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)

        for (i in 0 until childCount) {
            getChildAt(i).isEnabled = enabled
        }
    }

    override fun onSaveInstanceState(): Parcelable? {
        val savedState = SavedState(super.onSaveInstanceState())

        savedState.color = colorCenter.getColor()

        return savedState
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state is SavedState) {
            super.onRestoreInstanceState(state.superState)

            setColor(state.color)

        } else {
            super.onRestoreInstanceState(state)
        }
    }

    private fun addViewInternal(child: View) {
        isAddingInternal = true
        super.addView(child)
        isAddingInternal = false
    }

    override fun addView(child: View?) {
        if (isAddingInternal) {
            super.addView(child)
        }
    }

    override fun addView(child: View?, index: Int) {
        if (isAddingInternal) {
            super.addView(child, index)
        }
    }

    override fun addView(child: View?, width: Int, height: Int) {
        if (isAddingInternal) {
            super.addView(child, width, height)
        }
    }

    override fun addView(child: View?, params: LayoutParams?) {
        if (isAddingInternal) {
            super.addView(child, params)
        }
    }

    override fun addView(child: View?, index: Int, params: LayoutParams?) {
        if (isAddingInternal) {
            super.addView(child, index, params)
        }
    }

    override fun removeView(view: View?) {
    }

    override fun removeViewAt(index: Int) {
    }

    /**
     * set current picked color
     *
     * 设置选中颜色值
     *
     * @param color 颜色值
     */
    fun setColor(color: Int) {
        colorCenter.setColor(color, true)
    }

    /**
     * resolve touch conflict with it's parent and ancestors<br></br><br></br>
     *
     * if disallow is set to be true, [ColorPickerView]  will call
     * it's [parent][ViewParent.getParent]'s
     * [ViewParent.requestDisallowInterceptTouchEvent]
     * when received [MotionEvent.ACTION_DOWN] event in
     * [ColorPickerView.onTouchEvent]
     * to disallow it's parent and ancestors to intercept touch event,
     * and restore when received [MotionEvent.ACTION_UP] event
     * <br></br><br></br>
     * 解决[ColorPickerView]与其父View或祖先View的触摸事件冲突<br></br><br></br>
     *
     * disallow设置为true时，在[ColorPickerView.onTouchEvent]
     * 方法中接收到[MotionEvent.ACTION_DOWN] 触摸事件时会调用
     * [父View][ViewParent.getParent]的
     * [ViewParent.requestDisallowInterceptTouchEvent]方法来紧张父View
     * 及祖先View拦截触摸事件，并在收到[MotionEvent.ACTION_UP]触摸事件时恢复
     *
     * @param disallow whether to disallow it's parent and ancestors to intercept touch event
     *                 是否禁止父View或祖先View拦截触摸事件冲突
     */
    fun setDisallowInterceptTouchEvent(disallow: Boolean) {
        this.disallowInterceptTouchEvent = disallow
    }

    /**
     * set custom palette painter
     *
     * 设置自定义调色板绘制器
     *
     * @param painter custom palette painter
     *                       调色板绘制器
     */
    fun setPalettePainter(painter: IPalettePainter?) = apply { paletteView.painter = painter }

    /**
     * get custom palette painter
     */
    fun getPalettePainter() = paletteView.painter

    /**
     * set custom brightness painter
     */
    fun setBrightnessPainter(painter: IBrightnessPainter?) = apply { brightnessView.painter = painter }

    /**
     * get custom brightness painter
     */
    fun getBrightnessPainter() = brightnessView.painter

    /**
     * set custom alpha painter
     */
    fun setAlphaPainter(painter: IAlphaPainter?) = apply { alphaView.painter = painter }

    /**
     * get custom alpha painter
     */
    fun getAlphaPainter() = alphaView.painter

    override fun subscribe(observer: ColorObserver) {
        colorCenter.subscribe(observer)
    }

    override fun unsubscribe(observer: ColorObserver) {
        colorCenter.unsubscribe(observer)
    }

    override fun getColor() = colorCenter.getColor()

    private fun layout(view: View, viewGravity: Gravity, offsetX: Int, offsetY: Int) {
        val gravity = Gravity.calibrate(viewGravity)
        val w = view.measuredWidth
        val h = view.measuredHeight

        val left = when (gravity) {
            Gravity.LEFT_TOP, Gravity.LEFT_CENTER, Gravity.LEFT_BOTTOM -> 0
            Gravity.RIGHT_TOP, Gravity.RIGHT_CENTER, Gravity.RIGHT_BOTTOM -> measuredWidth - w
            else -> (measuredWidth - w) / 2
        } + offsetX

        val top = when (gravity) {
            Gravity.LEFT_TOP, Gravity.CENTER_TOP, Gravity.RIGHT_TOP -> 0
            Gravity.LEFT_BOTTOM, Gravity.CENTER_BOTTOM, Gravity.RIGHT_BOTTOM -> measuredHeight - h
            else -> (measuredHeight - h) / 2
        } + offsetY

        view.layout(
                left,
                top,
                left + w,
                top + h
        )
    }
}